### PR TITLE
Add CMS Implimentation and support of Windows Autheticode Code Signing 

### PR DESCRIPTION
**Pull request by RRRadicalEdward**
References:
- http://download.microsoft.com/download/9/c/5/9c5b2167-8017-4bae-9fde-d599bac8184a/authenticode_pe.docx - Windows Authenticode Portable Executable Signature Format
- https://datatracker.ietf.org/doc/html/rfc5652 - CMS RFC

### CODE REVIEW

File: picky-asn1-x509/src/pkcs7/content_info.rs [hunk 1]
```diff
@@ -0,0 +1,531 @@
...
+/// ``` not_rust
+/// SpcIndirectDataContent ::= SEQUENCE {
+///     data                    SpcAttributeTypeAndOptionalValue,
+///     messageDigest           DigestInfo
+/// }
+///
+/// SpcAttributeTypeAndOptionalValue ::= SEQUENCE {
+///     type                    ObjectID,
+///     value                   [0] EXPLICIT ANY OPTIONAL
+/// }
```
Comment (CBenoit) [inline picky-asn1-x509/src/pkcs7/content_info.rs:135]: This should be the docstring of `SpcAttributeAndOptionalValue`

File: picky-asn1-x509/src/pkcs7/content_info.rs [hunk 2]
```diff
@@ -0,0 +1,531 @@
...
+///     includeResources            (0),
+///     includeDebugInfo            (1),
+///     includeImportAddressTable   (2)
+/// }
+///
+/// SpcLink ::= CHOICE {
+///     url                     [0] IMPLICIT IA5STRING,
+///     moniker                 [1] IMPLICIT SpcSerializedObject,
+///     file                    [2] EXPLICIT SpcString
+/// } --#public--
```
Comment (CBenoit) [inline picky-asn1-x509/src/pkcs7/content_info.rs:211]: Put in `SpcLink`'s docstring

File: picky-asn1-x509/src/pkcs7/content_info.rs [hunk 3]
```diff
@@ -0,0 +1,531 @@
...
+///
+/// SpcString ::= CHOICE {
+///     unicode                 [0] IMPLICIT BMPSTRING,
+///     ascii                   [1] IMPLICIT IA5STRING
+/// }
```
Comment (CBenoit) [inline picky-asn1-x509/src/pkcs7/content_info.rs:216]: In `SpcString`'s docstring

File: picky-asn1-x509/src/pkcs7/content_info.rs [hunk 4]
```diff
@@ -0,0 +1,531 @@
...
+/// ``` not_rust
+/// SpcPeImageData ::= SEQUENCE {
+///    flags                   SpcPeImageFlags DEFAULT { includeResources },
+///    file                    SpcLink
+/// }
+/// SpcPeImageFlags ::= BIT STRING {
+///     includeResources            (0),
+///     includeDebugInfo            (1),
+///     includeImportAddressTable   (2)
+/// }
```
Comment (CBenoit) [inline picky-asn1-x509/src/pkcs7/content_info.rs:205]: In `SpcPeImageFlags`'s docstring.

File: picky-asn1-x509/src/pkcs7/crls.rs [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+/// RevocationInfoChoices ::= SET OF RevocationInfoChoice
+///
+/// RevocationInfoChoice ::= CHOICE {
+///    crl CertificateList,
+///    other [1] IMPLICIT OtherRevocationInfoFormat }
+///
+/// OtherRevocationInfoFormat ::= SEQUENCE {
+///    otherRevInfoFormat OBJECT IDENTIFIER,
+///    otherRevInfo ANY DEFINED BY otherRevInfoFormat }
+/// ```
```
Comment (CBenoit) [inline picky-asn1-x509/src/pkcs7/crls.rs:20]: Same as above

**Review by CBenoit — CHANGES_REQUESTED** [review]
PKCS7 implementation looks good, but I have a few change requests API-wise and others details.

File: picky/src/x509/pkcs7.rs [hunk 6]
```diff
@@ -0,0 +1,461 @@
...
+            digest_algorithm: DigestAlgorithmIdentifier(digest_algorithm.clone()),
+            signed_attrs: Attributes(authenticated_attributes).into(),
+            signature_algorithm: SignatureAlgorithmIdentifier(AlgorithmIdentifier::new_rsa_encryption()),
+            signature: encrypted_digest,
+        };
+
+        // certificates contains the signer certificate and any intermediate certificates,
+        // but typically does not contain the root certificate
+
+        let certificates = CertificateSet(certificates.0.into_iter().take(2).collect::<Vec<Certificate>>());
```
[the remark this answers is not in the document]
Comment (CBenoit) [reply picky/src/x509/pkcs7.rs:208]: This is good as is. :+1:

File: picky-asn1/src/restricted_string.rs [hunk 7]
```diff
@@ -247,6 +247,26 @@ impl CharSet for IA5CharSet {
...
+            return false;
+        }
+
+        let buffer = data
+            .chunks_exact(2)
+            .into_iter()
+            .map(|elem| u16::from_be_bytes([elem[1], elem[0]]))
+            .collect::<Vec<u16>>();
+
+        String::from_utf16(buffer.as_ref()).is_ok()
```
[the remark this answers is not in the document]
Comment (RRRadicalEdward) [reply picky-asn1/src/restricted_string.rs]: @awakecoding @CBenoit  
_char::decode_utf16(...)_ is stable from Rust 1.52, but CI cheсks are falling with the following: 
error Compiling picky-asn1 v0.3.1 (C:\__w\9\s\picky-asn1)
error[E0658]: use of unstable library feature 'assoc_char_funcs': recently added
   --> picky-asn1\src\restricted_string.rs:265:9
    |
265 |         char::decode_utf16(u16_it).all(|c| matches!(c, Ok(_)))
    |         ^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #71763 <https://github.com/rust-lang/rust/issues/71763> for more information

error: aborting due to previous error

For more information about this error, try `rustc --explain E0658`.
error: could not compile `picky-asn1`
I guess there is a lower version of Rust than 1.52 in the _devolutions/waykbuilder:linux_ and _devolutions/waykbuilder:vstools2k19_ . Could you upgrade the version, or should I use _String::from_utf16_ instead?

[the remark this answers is not in the document]
Comment (CBenoit) [reply picky-asn1/src/restricted_string.rs]: Oops, I didn't realize it got stabilized recently. I'll update wayk-builders so it builds, no problem.

File: picky-asn1-x509/src/pkcs7/crls.rs [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+///         revokedCertificates     SEQUENCE OF SEQUENCE  {
+///              userCertificate         CertificateSerialNumber,
+///              revocationDate          Time,
+///              crlEntryExtensions      Extensions OPTIONAL
+///                                       -- if present, version MUST be v2
+///                                   }  OPTIONAL,
+///         crlExtensions           [0]  EXPLICIT Extensions OPTIONAL
+///                                       -- if present, version MUST be v2
+///                                 }
+/// ```
```
Comment (CBenoit) [inline picky-asn1-x509/src/pkcs7/crls.rs:129]: Sorry, maybe I missed in the first review, but this one should be in the docstring of `TBSCertList`

File: picky-asn1-x509/src/pkcs7/signed_data.rs [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+///         version CMSVersion,
+///         digestAlgorithms DigestAlgorithmIdentifiers,
+///         encapContentInfo EncapsulatedContentInfo,
+///         certificates [0] IMPLICIT CertificateSet OPTIONAL,
+///         crls [1] IMPLICIT RevocationInfoChoices OPTIONAL,
+///         signerInfos SignerInfos }
+///
+///       DigestAlgorithmIdentifiers ::= SET OF DigestAlgorithmIdentifier
+///
+///       SignerInfos ::= SET OF SignerInfo
```
Comment (CBenoit) [inline picky-asn1-x509/src/pkcs7/signed_data.rs:23]: Same as above

File: picky-asn1-x509/src/pkcs7/signer_info.rs [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+///
+/// SignerIdentifier ::= CHOICE {
+///          issuerAndSerialNumber IssuerAndSerialNumber,
+///          subjectKeyIdentifier [0] SubjectKeyIdentifier }
+///
+/// SignedAttributes ::= SET SIZE (1..MAX) OF Attribute
+///
+/// UnsignedAttributes ::= SET SIZE (1..MAX) OF Attribute
+///
+/// SignatureValue ::= OCTET STRING
```
Comment (CBenoit) [inline picky-asn1-x509/src/pkcs7/signer_info.rs:28]: Same as above (if we define a type)

File: picky/src/hash.rs [hunk 11]
```diff
@@ -29,6 +47,39 @@ impl From<HashAlgorithm> for rsa::Hash {
         }
     }
 }
+impl TryFrom<HashAlgorithm> for SHAVariant {
```
Comment (CBenoit) [inline picky/src/hash.rs:50]:
```suggestion

impl TryFrom<HashAlgorithm> for SHAVariant {
```

File: picky/src/x509/mod.rs [hunk 12]
```diff
@@ -126,6 +126,11 @@
 //!# Ok(())
 //!# }
 //! ```
+#[cfg(feature = "pkcs7")]
```
Comment (CBenoit) [inline picky/src/x509/mod.rs:129]:
```suggestion

#[cfg(feature = "pkcs7")]
```

File: picky/src/pem.rs [hunk 13]
```diff
@@ -256,6 +256,8 @@ mod tests {
         pretty_assertions::assert_eq!(pem_from_reader, pem_from_str);
     }
 
+    #[cfg(not(windows))]
+    // This test should not run on Windows. writeln! add `/r` ending character to Pem in String format on Windows targets.
```
Comment (CBenoit) [inline picky/src/pem.rs:260]:
```suggestion
    // This test should not run on Windows. writeln! add `/r` ending character to Pem in String format on Windows targets.
    #[cfg(not(windows))]
```

File: picky/src/x509/pkcs7.rs [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+use picky_asn1_der::Asn1DerError;
+use picky_asn1_x509::algorithm_identifier::UnsupportedAlgorithmError;
+use picky_asn1_x509::pkcs7::Pkcs7Certificate;
+
+use thiserror::Error;
+
+type Pkcs7Result<T> = Result<T, Pkcs7Error>;
+
+pub const AUTHENTICODE_ATTRIBUTES_COUNT: usize = 3;
+
```
Comment (CBenoit) [inline picky/src/x509/pkcs7.rs:15]: We probably don't need that at all? Is that something you use in consumer code?

```suggestion
```

File: picky/src/x509/pkcs7.rs [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+    #[error(transparent)]
+    UnsupportedHashAlgorithmError(UnsupportedHashAlgorithmError),
+    #[error(transparent)]
+    UnsupportedAlgorithmError(UnsupportedAlgorithmError),
+    #[error("The signing certificate must contain the extended key usage (EKU) value for code signing")]
+    NoEKUCodeSigning,
+    #[error("Certificates must contain at least Leaf and Intermediate certificates, but got no certificates")]
+    NoCertificates,
+}
+const PKCS7_PEM_LABEL: &str = "PKCS7";
```
Comment (CBenoit) [inline picky/src/x509/pkcs7.rs:37]:
```suggestion

const PKCS7_PEM_LABEL: &str = "PKCS7";
```

**Review by CBenoit — CHANGES_REQUESTED** [review]
Thank you! Minor changes are required. As for the CI, I sent a PR to update the rust toolchain, no action is required on your part.

File: picky/src/x509/pkcs7.rs (hunk 14 above)
Comment (RRRadicalEdward) [reply picky/src/x509/pkcs7.rs:15]: Yes, We don't need it. I forget to remove it.

**Comment by RRRadicalEdward** [comment]
@CBenoit I resolved the latest issues. Is there still something that blocks the merging? 🙂

**Review by CBenoit — APPROVED** [review]
Thank you, this looks good to me! :smile:

**Comment by RRRadicalEdward** [comment]
> 
> 
> Thank you, this looks good to me! 😄
> 
> (I just moved a docstring around, since it's not much I went ahead and pushed directly)
Thank you for the review. picky is awesome.

**Review by awakecoding — APPROVED** [review]
LGTM

**Comment by RRRadicalEdward** [comment]
@CBenoit Could you merge it? Seems like I cannot merge because of falling CI checks.

**Comment by awakecoding** [comment]
@RRRadicalEdward Benoit has been working on fixing the CI with the Rust update, it should be updated soon

**Comment by CBenoit** [comment]
CI is green!
@RRRadicalEdward I'll publish a new version of picky & ci :slightly_smiling_face:

**Pull request merged**